pandas/_libs/src/inference.pyx

```diff
@@ -593,190 +598,268 @@ cdef inline bint is_timedelta(object o):
     return PyDelta_Check(o) or util.is_timedelta64_object(o)
 
 
-cpdef bint is_bool_array(ndarray values):
+cdef class Validator:
+
     cdef:
-        Py_ssize_t i, n = len(values)
-        ndarray[object] objbuf
+        Py_ssize_t n
+        object dtype
+        bint skipna
 
-    if issubclass(values.dtype.type, np.bool_):
-        return True
-    elif values.dtype == np.object_:
-        objbuf = values
+    def __cinit__(self, Py_ssize_t n, object dtype=None, bint skipna=False):
+        self.n = n
+        self.dtype = dtype if dtype is not None else np.dtype(np.object_)
+        self.skipna = skipna
+
+    cdef bint validate(self, object[:] values):
+        if not self.n:
+            return False
 
-        if n == 0:
+        if self.is_array_typed():
+            return True
+        elif self.dtype.type == np.object_:
+            if self.skipna:
+                return self._validate_skipna(values)
+            else:
+                return self._validate(values)
+        else:
             return False
 
+    @cython.wraparound(False)
+    @cython.boundscheck(False)
+    cdef bint _validate(self, object[:] values):
+        cdef:
+            Py_ssize_t i
+            Py_ssize_t n = self.n
+
         for i in range(n):
-            if not util.is_bool_object(objbuf[i]):
+            if not self.is_valid(values[i]):
                 return False
-        return True
-    else:
+
+        return self.finalize()
+
+    @cython.wraparound(False)
+    @cython.boundscheck(False)
+    cdef bint _validate_skipna(self, object[:] values):
+        cdef:
+            Py_ssize_t i
+            Py_ssize_t n = self.n
+
+        for i in range(n):
+            if not self.is_valid_skipna(values[i]):
+                return False
+
+        return self.finalize()
+
+    cdef bint is_valid(self, object value) except -1:
+        return self.is_value_typed(value)
+
```

Comment:
I think I might be able to, didn't try. I would also except the compiler to be pretty smart here. `inline` shouldn't have much affect nowadays, since the compiler will ignore it if it's unsafe to do.